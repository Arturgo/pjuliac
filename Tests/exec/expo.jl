
# exponentiation rapide

function expo(x, n)
    # n >= 1
    if n == 1
        x
    else
        y = expo(x, div(n, 2))
        y = mult(y, y)
        if n % 2 == 0 y else mult(x, y) end
    end
end

function mult(x::Int64, y::Int64) x * y end

println(expo(2, 1))
println(expo(2, 10))

# fibonacci en temps O(log n)

struct Mat22 a11; a12; a21; a22 end

function mult(x::Mat22, y::Mat22)
    Mat22(x.a11 * y.a11 + x.a12 * y.a21, x.a11 * y.a12 + x.a12 * y.a22,
          x.a21 * y.a11 + x.a22 * y.a21, x.a21 * y.a12 + x.a22 * y.a22)
end

fibM = Mat22(1, 1, 1, 0)
function fib(n)
    m = expo(fibM, n)
    m.a12
end

for n = 1:10 println(fib(n)) end

println(fib(1000000000))  # avec un débordement, bien sûr, mais instantané

