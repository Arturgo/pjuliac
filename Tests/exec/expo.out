2
1024
1
1
2
3
5
8
13
21
34
55
3311503426941990459
